docs/src/proposals/timely-vote-credits.md

```diff
@@ -0,0 +1,144 @@
+slots, every additional slot of latency would result in a 10% reduction in
+vote credits earned as each subsequent slot earns 1 credit less out of a
+maximum possible 10 credits.
+
+Again, historical voting data was analyzed over many epochs and the conclusion
+drawn was that a maximum credits of 10 is the largest value that can be used
+and still have a noticeable effect on known laggers.  Values higher than that
+result in such a small penalty for each slot of lagging that intentional
+lagging is still too profitable.  Lower values are even more punishing to
+intentional lagging; but an attempt has been made to conservatively choose the
+highest value that produces noticeable results.
+
+The selection of these values is partially documented here:
+
+https://www.shinobi-systems.com/timely_voting_proposal
+
+The above document is somewhat out of date with more recent analysis, which
+occurred in this github issue:
+
+https://github.com/solana-labs/solana/issues/19002
+
+To summarize the findings of these documents: analysis over many epochs showed
+that almost all validators from all regions have an average vote latency of 1
+slot or less.  The validators with higher average latency are either known
+laggers, or are not representative of their region since many other validators
+in the same region achieve low latency voting.  With a maximum vote credit of
+10, there is almost no change in vote credits earned relative to the highest vote
+earner by the majority of validators, aside from a general uplift of about 0.4%.
+Additionally, data centers were analyzed to ensure that there aren't regions of
+the world that would be adversely affected, and none were found.
+
+
+### Method of Implementation
+
+When a Vote or VoteStateUpdate instruction is received by a validator, it will
+use the Clock sysvar to identify the slot in which that instruction has
+landed.  For any newly voted on slot within that Vote or VoteStateUpdate
+transaction, the validator will record the vote latency of that slot as
+(voted_in_slot - voted_on_slot).
+
+These vote latencies will be stored a new vector of u8 latency values appended
+to the end of the VoteState.  VoteState currently has ~200 bytes of free space
+at the end that is unused, so this new vector of u8 values should easily fit
+within this available space.  Because VoteState is an ABI frozen structure,
+utilizing the mechanisms for updating frozen ABI will be required, which will
+complicate the change.  Furthermore, because VoteState is embedded in the
+Tower data structure and it is frozen ABI as well, updates to the frozen ABI
+mechanisms for Tower will be needed also.  These are almost entirely
+mechanical changes though, that involve ensuring that older versions of these
+data structures can be updated to the new version as they are read in, and the
+new version written out when the data structure is next persisted.
+
+The credits to award for a rooted slot will be calculated using the latency
```

Comment:
could you elaborate a little bit on which latencies will be used? I assume it's not from the rooted slot itself as there are no votes for that slot present. When a slot gets rooted are we actually paying out the votes for the previous slots?

Reply:
When a vote is on a slot A, the earliest slot that that vote can land in is slot A + 1.  That is given latency 0.  Therefore latency is ((slot vote landed in - 1) - (slot vote was for)).

Latencies 0 through 2 are awarded maximum credits, then credits reduce 1 per extra slot of latency after that.

It is true that sometimes there will be skipped blocks so it may be impossible to get full credits as this may push latencies very high.  But this affects all validators equally, so relative credits earning performance isn't affected.

When a slot gets rooted, it carries with it the latency of the vote that added it to Lockouts; and that latency is used to determine the credits to pay.

Therefore: latency is recorded when a vote first lands; but the credits for that slot are not awarded until the slot gets rooted, at which point the latency that was recorded is used to compute the credits to award for that newly rooted slot.

Do you think I need to add more text to the doc to explain this, or is it clear already?

Reply:
Got it that clears up my confusion. 
I think if you added something like your second to last paragraph "latency is recorded when a vote first lands; but the credits for that slot are not awarded until the slot gets rooted, at which point the latency that was recorded is used to compute the credits to award for that newly rooted slot." that would be helpful